### PR TITLE
Add min_delta to early_stopping's _stop_if_no_metric_improvement

### PR DESCRIPTION
**Pull request by jquadrino**
I added a min_delta argument that allows for more specific metric improvement criteria.

### CODE REVIEW

**Comment by guillaumekln** [comment]
What is blocking this PR? This is a useful feature.

**Comment by karmel** [comment]
@rchao to review whether this makes sense/can be supported. This should minimally come with tests, though-- @jquadrino , can you add some tests that cover this functionality?

**Comment by rchao** [comment]
Please apply this change to estimator/python/estimator/early_stopping.py. Thanks!

**Comment by rthadur** [comment]
> Please apply this change to estimator/python/estimator/early_stopping.py. Thanks!

@jquadrino gentle ping

**Comment by jquadrino** [comment]
@rthadur @rchao I moved the logic to the appropriate module. Please review, adding the unit tests in a moment.

File: tensorflow_estimator/python/estimator/early_stopping.py [hunk 1]
```diff
@@ -177,13 +181,18 @@ def read_eval_metrics(eval_dir):
...
                                     higher_is_better, eval_dir, min_steps,
-                                    run_every_secs, run_every_steps):
+                                    min_delta, run_every_secs, run_every_steps):
   """Creates early-stopping hook to stop training if threshold is crossed."""
 
   if eval_dir is None:
     eval_dir = estimator.eval_dir()
 
-  is_lhs_better = operator.gt if higher_is_better else operator.lt
+  def is_lhs_better(lhs_val, val):
```
Comment (rchao) [inline tensorflow_estimator/python/estimator/early_stopping.py]: I would rename the function now that it's not "only" better, but better by min_delta

File: tensorflow_estimator/python/estimator/early_stopping.py [hunk 2]
```diff
@@ -130,6 +131,8 @@ def stop_if_higher_hook(estimator,
       default, `estimator.eval_dir()` will be used.
     min_steps: `int`, stop is never requested if global step is less than this
       value. Defaults to 0.
+    min_delta: `float`, minimum required change in the given metric to determine
```
Comment (rchao) [inline tensorflow_estimator/python/estimator/early_stopping.py]: Is it accepted to pass in negative `min_delta`?

**Comment by jquadrino** [comment]
@rchao changes made! Let me know how I should handle min_delta value validation.

File: tensorflow_estimator/python/estimator/early_stopping.py (hunk 2 above)
Comment (jquadrino) [reply tensorflow_estimator/python/estimator/early_stopping.py]: No it is not, I can enforce this by grabbing the absolute value or asserting > 0.

Comment (rchao) [reply tensorflow_estimator/python/estimator/early_stopping.py]: How does it sound if we raise `ValueError` if `min_delta < 0` and add this info to the docstring?

Comment (jquadrino) [reply tensorflow_estimator/python/estimator/early_stopping.py]: Sounds good, I added the validation to the `stop_if_higher_hook` function.

**Comment by jquadrino** [comment]
@rchao I applied the changes to the remaining functions at head.

File: tensorflow_estimator/python/estimator/early_stopping_test.py [hunk 3]
```diff
@@ -135,14 +135,17 @@ def run_session(self, hooks, should_stop):
         mon_sess.run(no_op)
         self.assertEqual(mon_sess.should_stop(), should_stop)
 
-  @parameterized.parameters((0.8, 0, False), (0.6, 4000, False), (0.6, 0, True))
-  def test_stop_if_higher_hook(self, threshold, min_steps, should_stop):
+  @parameterized.parameters((0.8, 0, 0, False), (0.8, 0, 0.05, True),
+                            (0.6, 4000, 0, False), (0.6, 0, 0, True))
+  def test_stop_if_higher_hook(self, threshold, min_steps, min_delta,
```
Comment (rchao) [inline tensorflow_estimator/python/estimator/early_stopping_test.py]: Can we also revise the below tests?

**Comment by rchao** [comment]
> @rchao I applied the changes to the remaining functions at head.

Thanks for the change! Just a couple of nits to address then I think this is good to go.

**Comment by rthadur** [comment]
@jquadrino can you please check review comments

**Comment by jquadrino** [comment]
@rchao @rthadur Addressed comments, ready for review.

**Comment by karmel** [comment]
Is this change necessary? Why can't you subclass/write your own hook in this case? We are wary of adding additional complexity to this particular set of hooks/APIs.

**Comment by rthadur** [comment]
@jquadrino can you please latest comments from Karmel ?

**Comment by jquadrino** [comment]
Yes, those comments seem late but I understand its an evolving project. Original intent was to mimic keras' EarlyStopping to better be able to interchange keras and tf-estimator frameworks. I will go ahead and close this.
